src/main/java/com/ibm/cloud/cloudant/kafka/connect/CloudantSinkTask.java

```diff
@@ -101,13 +107,51 @@ public void start(Map<String, String> props) {
 
 	@Override
 	public void flush(Map<TopicPartition, org.apache.kafka.clients.consumer.OffsetAndMetadata> offsets) {
-		LOG.debug("Flushing output stream for {" + config.getString(InterfaceConst.URL) + "}");
+
+		System.out.println(Arrays.toString(Thread.currentThread().getStackTrace()));
+
+		List<Map<String, Object>> jsonArray = new LinkedList<>();
+
+		if (accumulatedSinkRecords != null) {
+			// Note: _rev is preserved
+			accumulatedSinkRecords.stream()
+					.map(mapper) // Convert ConnectRecord to Map
+					.sequential() // Avoid concurrent access to jsonArray
+					.forEach(recordValueAsMap -> {
+						jsonArray.add(recordValueAsMap);
+					});
+		}
+
+
 		try {
-				JavaCloudantUtil.batchWrite(config.originalsStrings(), jsonArray);
-				LOG.info("Committed " + jsonArray.size() + " documents to -> " + config.getString(InterfaceConst.URL));
+			LOG.info(String.format("Calling batchWrite with %d documents to %s", jsonArray.size(), config.getString(InterfaceConst.URL)));
+			List<DocumentResult> writeResults = JavaCloudantUtil.batchWrite(config.originalsStrings(), jsonArray);
+			boolean ok = writeResults.stream().allMatch(DocumentResult::isOk);
+			if (!ok) {
+				LOG.error("Failed to write some documents");
+				for (int i=0; i<writeResults.size(); i++) {
+					// TODO - is checking isOk sufficient?
+					// TODO - are bulk doc results guaranteed to be in same order?
+					// TODO - think of a better exception to raise here
+					if (!writeResults.get(i).isOk()) {
+						reporter.report(accumulatedSinkRecords.get(i), new RuntimeException("not ok"));
+					}
+				}
+			}
+		} catch (RuntimeException re) {
+			LOG.error(String.format("Exception thrown when trying to write documents: %s", re.getMessage()));
+			if (reporter != null) {
+				if (accumulatedSinkRecords != null) {
+					// all failed
+					for (SinkRecord r : accumulatedSinkRecords) {
+						reporter.report(r, re);
+					}
+				}
+			} else {
+				throw new ConnectException("Exception thrown when trying to write documents", re);
+			}
 		} finally {
-			// Release memory (regardless if documents got committed or not)
-			jsonArray = new LinkedList<>();
+			accumulatedSinkRecords = null;
```

Comment:
I'm a little uneasy about this, what if we only processed half the list when an exception occurs, won't we "lose" the accumulated records? I think ultimately it will depend on whether the offsets for those records have been committed.

AIUI the offsets are "commited" after `put` is called, but further reading suggests that they are just committed on a cron, and `put` throwing an exception is used to determine what is [not] eligible and needs retrying. Now that we are not writing during `put` we may want to be a more careful about the offsets. For example, a 60 second commit timer seems inappropriate if it takes us 2 minutes to write a batch, because we'd commit offsets before we actually wrote the records.

I think we may need to override `preCommit` and have it "replace" `flush`, but return the record offsets that we've actually processed so Kafka knows what is eligible to commit. [This](https://stackoverflow.com/a/45202162) seems to be helpful information on this topic.

Reply:
I've stuck with overriding `flush` in 6a8928b , but the logic has changed to only clear the list if all was successful.

If we get an exception from batch write, then we re-throw it wrapped (this isn't necessary as the comments make clear, and if you think I should get rid of the wrapping, I'm happy to). This will ensure that the `accumulateSinkRecords` are re-tried the next time `flush` is called - I verified this by replacing the batch write with a stub which throws exceptions every time.